mockfactory/desi/imaging_maskbits.py

```diff
@@ -73,34 +74,54 @@ def get_brick_maskbits(maskbits_fn, ra, dec):
             maskbits = maskbits_img[coadd_y, coadd_x]
 
         else:
-            #raise ValueError
+            # raise ValueError
             # Sometimes we can have objects outside DR9 footprint:
             # remove these objects setting maskbits 0 (NPRIMARY pixel)
             maskbits = 2**0 * np.ones(ra.size, dtype=dtype)
 
         return maskbits
 
-    # If not empty, collect the information from brick
     ra, dec = np.asarray(ra), np.asarray(dec)
     maskbits = np.ones_like(ra, dtype='i2')
-    if ra.size != 0:
-        # Collect brickid
-        bricks = brick.Bricks()
-        cumsize = np.cumsum([0] + mpicomm.allgather(ra.size))[mpicomm.rank]
-        index = cumsize + np.arange(ra.size)
-        data = _dict_to_array({'ra': ra, 'dec': dec, 'brickname': bricks.brickname(ra, dec), 'brickid': bricks.brickid(ra, dec), 'maskbits': maskbits, 'index': index})
-        #data = _dict_to_array({'ra': ra, 'dec': dec, 'brickname': bricks.brickname(ra, dec), 'brickid': index[::-1], 'maskbits': maskbits, 'index': index})
-        import mpsort
-        mpsort.sort(data, orderby='brickid')
-
-        for brickname in np.unique(data['brickname']):
-            mask_brick = data['brickname'] == brickname
-            region = 'north' if bricks.brick_radec(data['ra'][mask_brick][0], data['dec'][mask_brick][0])[1] > 32.375 else 'south'
-            data['maskbits'][mask_brick] = get_brick_maskbits(maskbits_fn.format(region=region, brickname=brickname), data['ra'][mask_brick], data['dec'][mask_brick])
-
-        mpsort.sort(data, orderby='index')
-        assert np.all(data['index'] == index)
-        maskbits = data['maskbits']
+
+    # load bricks class
+    bricks = brick.Bricks()
+
+    # create unique identification as index column
+    cumsize = np.cumsum([0] + mpicomm.allgather(ra.size))[mpicomm.rank]
+    index = cumsize + np.arange(ra.size)
+    data = _dict_to_array({'ra': ra, 'dec': dec, 'brickname': bricks.brickname(ra, dec), 'brickid': bricks.brickid(ra, dec), 'maskbits': maskbits, 'index': index})
+
+    # since we want to parrallelize around the brickid
+    # we do not expect that the number of particles is the same as the input on each rank
+    # use out argument in mpsort.sort function
+    unique_brickname, brick_counts = np.unique(np.concatenate(mpicomm.allgather(data['brickid'])), return_counts=True)
```

Comment:
This will fail for large data['brickid'] arrays (>2 GB) and may blow up the memory (duplicating array on all ranks).
It should be possible to call np.unique() on each process, then take again np.unique(), only on rank 0, and perform the calculation of nbr_particles on rank 0.

Reply:
correct